### DISCUSSION TITLE
Chapter 1 - Exercise

### DISCUSSION THREAD

**Discussion by michaelshekasta** [huggingface/computer-vision-course#290] [01. Fundamentals]
Hey everyone,

I'm excited to introduce a new exercise for this unit that I believe will be both engaging and beneficial. My concept revolves around the idea of finding a picture within another picture, offering a practical application of the concepts we've been exploring.

I've outlined the details in the draft PR, which you can find here: [Link to PR](https://github.com/johko/computer-vision-course/pull/289)

Looking forward to your feedback and collaboration on this!

Michael

**Comment by michaelshekasta** [comment]
@bellabf

**Comment by bellabf** [comment]
Hi, Michael! I think this would be cool but maybe to keep things more cohesive we could do it with [SIFT and feature matching(https://huggingface.co/learn/computer-vision-course/unit1/feature-extraction/feature-matching)] instead of NN? It would be fit better content wise.

**Reply by michaelshekasta** [reply]
@bellabf 
Thank you for your response. It seems there may have been a misunderstanding. To tackle the exercise without ANN, you should use the sift technique.

**Comment by bellabf** [comment]
> @bellabf Thank you for your response. It seems there may have been a misunderstanding. To tackle the exercise without ANN, you should use the sift technique.

Yes, sorry. I read the file quickly and didn't realize you said without NN. I think it might just be easier/clearer to say using the SIFT technique (it leaves less room for guessing).

 I would be interested to hear more about other people's perspective on adding exercises that are not graded, cc @mmhamdy @ATaylorAerospace @merveenoyan and also @johko (who might a take a bit longer to reply to this) considering there is also another discussion https://github.com/johko/computer-vision-course/discussions/295 on adding exercise to another unit 2, It might be worth it to discuss whether to take a broader effort and how they would be structured.

**Reply by ATaylorAerospace** [reply]
@bellabf One of the goals of the current version of the course was to have the exercises be graded in order to receive a certificate. But I remember @johko mentioning that over time as novel approaches get introduced in Computer Vision as a whole we may need to add onto the course.  

So with that being the case I would think that there may be new course topics and associated non graded exercises that may have to be added to the course.

**Reply by michaelshekasta** [reply]
@ATaylorAerospace @bellabf 
I'd like to propose an enhancement: incorporating self-practice exercises and assignments. Specifically, for the initial two units, we could develop assignments where participants are tasked with building models using a provided training set and then evaluating their performance through a leaderboard displayed via Gradio. What are your thoughts on this idea?

**Reply by mmhamdy** [reply]
Hi @michaelshekasta, thanks for suggesting the exercises. I think end-of-unit exercises are nice to have. They can help in measuring your understanding and what you took away from each unit. Other Hugging Face courses have them. But as @ATaylorAerospace mentioned, they need to be graded (not necessarily for certification) so that the reader can easily get feedback. Multiple-choice questions are an example of exercises that are really helpful in measuring your understanding and very easy to evaluate.

**Reply by michaelshekasta** [reply]
@mmhamdy @ATaylorAerospace @johko @bellabf @merveenoyan 
I plan to customize the exercise as follows: I'll curate a test dataset comprising numerous images and randomly crop portions from each image to create smaller ones. These pairs of original images and their cropped segments will then be uploaded to the Hugging Face dataset repository.

Subsequently, I intend to develop a Leaderboard using Gradio. This leaderboard will evaluate models by calculating the Mean Average Precision (MAP) between the ground truth and the model's outputs.
What are your thoughts on this approach?

**Reply by bellabf** [reply]
The exercise idea is pretty cool. For unit one, creating a leaderboard might be an overshoot. The unit is more introductory. The exercise proposed also use SIFT (which only has 5 parameters) so I think it makes this less interesting for leaderboard.  For other units, this might be better, since the models have more parameters and are more complex.

In the same line of thought, I think there are too many units (total of thirteen) to make a challenge with a leaderboard for every unit. At some point, I think it gets too overwhelming. I would think it is best to create maybe two, or max of three, for the whole course.

I agree with @mohammad-gh009 that the multiple choice quizzes provide a quick way to gain feedback and are also easy to create. [This idea is also implemented as an end-of-the chapter quiz in the NLP course, which we could take inspiration from. ](https://huggingface.co/learn/nlp-course/chapter1/10?fw=pt)

**Reply by mmhamdy** [reply]
@michaelshekasta that sounds cool! I'm excited about the leaderboard, but I agree with @bellabf in that having many leaderboards is an overkill. In my opinion, it would be nice to have simple exercises for each unit and one final leaderboard at the end of the course.

**Reply by michaelshekasta** [reply]
@mmhamdy @bellabf 
I get your perspective, but I sense there's more to consider here. Let's take a cue from the HF audio course, where they've structured it with four assignments graded via Gradio for easy score checking.

Given this, might it be beneficial for us to adopt a comparable approach? Particularly as we advance into chapters that necessitate more substantial resources such as GPUs. Personally, I prefer assignments with coding or hands-on tasks instead of regular questions.

What do you think?
@johko

**Reply by michaelshekasta** [reply]
P.S.  Here is the link for checking progress in the audio course:
https://huggingface.co/spaces/MariaK/Check-my-progress-Audio-Course

**Reply by bellabf** [reply]
This discussion rings a bell. The audio course uses the gradio to check the progress in order to get a certificate of completion.  I believe the criteria for the certificate of completion was made by the organization team. In my opinion, it all comes down to what it is easier for the person that has to issue the certificate of completion. 

One thing to keep in mind that is that exercises to check understanding, certificate of completion criteria, and leaderboard challenges serve different goals (which would learning, passing, and engagement, respectively). It might be a good idea to reflect about what it is that we want to accomplish by adding exercises to the unit.

**Reply by michaelshekasta** [reply]
@bellabf I totally agreed!
@mmhamdy @johko @merveenoyan What do you think?

**Reply by mmhamdy** [reply]
I agree. Leaderboards are more engaging and fun but this requires reevaluating our certification process (as @bellabf said), which I don't actually mind. But @johko and @merve would know better, so let's see what they think about that.

**Reply by johko** [reply]
To be fair - I don't know that much better :smile: But I do have an opinion on the whole thing.

I do like your proposal @michaelshekasta but it also sounds like a lot of work maintaining and creating an exercise like that for every unit and I'm not sure we can handle that in a good way

I also think it would be a good way to make the certification multi-step, similar to the audio course. Right now I feel like somebody could just take some model, do some fine-tuning, upload it and basically be done without having to look at the course too much.

I think we should come up with some key exercises that are taken into account for the overall certification. I don't know which ones that would be yet and we probably have to involve @merveenoyan in the whole planning, as she already mentioned that she wants to create a leaderboard.

For the other units it might be good enough to have a multiple choice quiz at the end (you know I like quizzes) and they are not that hard to create as you can see in this file: https://github.com/huggingface/audio-transformers-course/blob/main/chapters/en/chapter1/quiz.mdx?plain=1

tl;dr graded exercises yes, but not for every unit

**Comment by michaelshekasta** [comment]
@johko, thank you for your comment! I believe it's important to keep in mind that utilizing advanced units like VLAM models requires substantial GPU resources, which could pose challenges for students in the course.

I've outlined some assignment directions already:

Unit 1: Identify small images within larger images using SIFT.
Unit 2: Train a neural network on established datasets like MNIST.
As for labs, which serve as valuable exercises, for Unit 2, one idea is to compare a single CNN with numerous kernels against the approach of using two CNNs sequentially with fewer kernels each.
